### PR TITLE
feat: overrides & improvements to tool

### PR DESCRIPTION
**Pull request by xander-jones**
## Goal

* Improve usability of the `bugsnag-dsym-upload` tool, whilst appropriately erroring and warning users when dSYM _files_, or dSYM directories without DWARF debug data are found.
* The default behaviour of this tool is not changed in this PR.

## Design

Design approach discussed with Bugsnag team to ensure we're not making any breaking changes, whilst allowing further customization of options for users.

## Changeset

* Switch to use `find` rather than `glob` to catch all dSYMs regardless of directory depth.
* missing DWARF data: When a .dSYM directory is missing DWARF data show error, exit code non-zero, as before.
	* With `-m/--allow-missing-dwarf` option enabled, this becomes a warning, (exit code 0, if no other errors)
* 0b dSYM files: When looking for *.dSYM: find files _and_ directories ending “.dSYM”, 
	* if directory: continue upload as normal
    * if file: show suitable error (previously failed with "missing DWARF" error, exit code non zero.
    	* With `-e/--ignore-empty-dsym` option enabled, show warning instead, exit code 0.
* Fastlane action `upload_symbols_to_bugsnag.rb` has been updated to make use of these new option flags
* Always output errors and warnings from our CLI tool, not just in verbose mode, to help users debug when Fastlane spits out strange errors

## Testing

Automated tests were added to `features/dsym_upload.feature`, and `features/fastlane_dsym_upload.feature`. New fixtures have also been added to aid this testing.

Binary was also manually tested outside of Fastlane environment to ensure it was working as expected

### CODE REVIEW

File: bin/bugsnag-dsym-upload [hunk 1]
```diff
@@ -13,18 +13,22 @@
...
-    echo "--upload-server URL The server receiving dSYM files. Set this value if"
-    echo "                    using an on-premise Bugsnag installation"
-    echo "--project-root DIR  The root directory of the project. This will help to"
-    echo "                    group error reports by project"
-    echo "dSYMS_PATH          A directory or .zip file containing *.dSYM bundles to"
-    echo "                    upload"
+    echo "-h, --help                 Displays this message"
+    echo "-v, --verbose              Print verbose logging output during execution"
+    echo "-m, --allow-missing-dwarf  Throw warnings instead of errors when a dSYM with missing"
+    echo "                           DWARF data is found"
```
Comment (tomlongridge) [inline bin/bugsnag-dsym-upload:19]: These two options both feel like "ignore" options, so I would suggest we have `--ignore-missing-dwarf` for consistency.

File: bin/bugsnag-dsym-upload [hunk 2]
```diff
@@ -13,18 +13,22 @@
...
+    echo "-e, --ignore-empty-dsym    Throw warnings instead of errors when a *.dSYM file is found"
+    echo "                           rather than the expected *.dSYM directory"
```
Comment (tomlongridge) [inline bin/bugsnag-dsym-upload:21]: I'm not convinced the short options add anything here – it's not like it's an intuitive character nor an option that people are going to need to type in again and again. I suggest we remove it for consistency with the other non-standard options.

Comment (tomlongridge) [reply bin/bugsnag-dsym-upload:21]: I'd also suggest these should go as the last options in the usage text as they are a bit more niche than the others.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,12 @@
+## TBC (TBC)
+
+### Enhancements
+
+* Files found with `Dir.glob` are now sorted alphabetically so that this action is deterministic accross OS's
+* Adds -m/--allow-missing-dwarf option to throw warnings instead of errors when a dSYM with missing DWARF data is found
+* Adds -e/--ignore-empty-dsym option to throw warnings instead of errors when a *.dSYM file is found rather than the expected *.dSYM directory"
+  [#TBD](TBD)
```
Comment (tomlongridge) [inline CHANGELOG.md:8]:
```suggestion
  [#60](https://github.com/bugsnag/bugsnag-dsym-upload/pull/60)
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,12 @@
+## TBC (TBC)
+
+### Enhancements
+
+* Files found with `Dir.glob` are now sorted alphabetically so that this action is deterministic accross OS's
+* Adds -m/--allow-missing-dwarf option to throw warnings instead of errors when a dSYM with missing DWARF data is found
+* Adds -e/--ignore-empty-dsym option to throw warnings instead of errors when a *.dSYM file is found rather than the expected *.dSYM directory"
```
Comment (tomlongridge) [inline CHANGELOG.md:7]: Can we cite some web references or give examples of when users may need these? Or put it in the PR description? Just want to make sure we keep the rationale.

File: CHANGELOG.md (hunk 3 above)
Comment (tomlongridge) [reply CHANGELOG.md:8]: Are there also some GH issue numbers to reference here (and/or in the PR description)?

**Review by tomlongridge — COMMENTED** [review]
Left some comments on the non-technical side. I'll defer to @nickdowell for technical review.

File: bin/bugsnag-dsym-upload (hunk 2 above)
Comment (xander-jones) [reply bin/bugsnag-dsym-upload]: allow changed for ignore, moved to bottom of options list: [`aa238a2` (#60)](https://github.com/bugsnag/bugsnag-dsym-upload/pull/60/commits/aa238a217d45bb9e48ae81022ff0d3f5623f89e9)

File: bin/bugsnag-dsym-upload (hunk 1 above)
Comment (xander-jones) [reply bin/bugsnag-dsym-upload]: allow changed for ignore [`aa238a2` (#60)](https://github.com/bugsnag/bugsnag-dsym-upload/pull/60/commits/aa238a217d45bb9e48ae81022ff0d3f5623f89e9)

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,13 @@
+## TBC (TBC)
+
+### Enhancements
+
+* Switch to use `find` rather than `glob` to catch all dSYMs regardless of directory depth (for AppCenter dSYM bundle support)
+* Files found with `Dir.glob` are now sorted alphabetically so that this action is deterministic accross OS's
+* Adds -m/--ignore-missing-dwarf option to throw warnings instead of errors when a dSYM with missing DWARF data is found
+* Adds -e/--ignore-empty-dsym option to throw warnings instead of errors when a *.dSYM file is found rather than the expected *.dSYM directory"
```
Comment (nickdowell) [inline CHANGELOG.md:8]: Short options need to be removed from here!

File: bin/bugsnag-dsym-upload [hunk 6]
```diff
@@ -65,6 +71,12 @@ while [[ $# -gt 0 ]]; do
         -v|--verbose)
             verbose=1
             shift;;
+        -m|--ignore-missing-dwarf)
+            ignore_missing_dwarf=1
+            shift;;
+        -e|--ignore-empty-dsym)
+            ignore_empty_dsym=1
+            shift;;
```
Comment (nickdowell) [inline bin/bugsnag-dsym-upload:79]: The short options are still here...?

**Review by nickdowell — CHANGES_REQUESTED** [review]

**Review by nickdowell — APPROVED** [review]

**Comment by xander-jones** [comment]
adding @kattrali as she has context as we discussed the move of the `man` pages

**Review by tomlongridge — APPROVED** [review]

**Review by kattrali — APPROVED** [review]

**Pull request merged**